[Security Solution] Implement prebuilt rules export test plan

maximpn opened the pull request:
**Addresses:** https://github.com/elastic/kibana/issues/202079
**Relates to:** https://github.com/elastic/kibana/pull/222796

## Summary

This PR implements Prebuilt Rules export [test plan](https://github.com/elastic/kibana/blob/main/x-pack/solutions/security/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/prebuilt_rule_export.md).

Existing tests were adjusted and extended to match the test plan.

## Caveats

The test plan describes a scenario to verify prebuilt rules export fails for 10K+ rules but this scenario is tricky to implement. Due to ES limitations on filtering more than 10K the majority of utility functions like `deleteAllRules()` fail. The proper implementation requires proper setup and cleanup to make sure the test doesn't block the testing workflow.

As the result of the mentioned complexities implementation of the test scenario for 10K+ rules is skipped in this PR.


<!--ONMERGE {"backportTargets":["8.18","8.19","9.0","9.1"]} ONMERGE-->

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

maximpn committed:
implement export prebuilt rules test plan

maximpn commented:
/ci

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/312522)
* Commit: eca01bb641386b49e9670aeffc113fa7b98f1eb3

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #312385](https://buildkite.com/elastic/kibana-pull-request/builds/312385) failed 197f629e3872c479966b184268aae0160a040545
* :broken_heart: [Build #312202](https://buildkite.com/elastic/kibana-pull-request/builds/312202) failed a51b4d6857ffd9f9ec6b59bfb902edfb30c9af4f

cc @maximpn

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312522","number":312522,"commit":"eca01bb641386b49e9670aeffc113fa7b98f1eb3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312385","number":312385,"commit":"197f629e3872c479966b184268aae0160a040545"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312202","number":312202,"commit":"a51b4d6857ffd9f9ec6b59bfb902edfb30c9af4f"}],"number":312522}
buildkite-pr-comment-->

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detection-rule-management (Team:Detection Rule Management)

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/313863)
* Commit: a86dea3a8b6182f0693015cd201b103af387d2ed

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #312777](https://buildkite.com/elastic/kibana-pull-request/builds/312777) succeeded 341917740e36cea21953f7735a564d11cc7fd134
* :broken_heart: [Build #312676](https://buildkite.com/elastic/kibana-pull-request/builds/312676) failed f6cdc52acc443524ac119cb65798d6b344892a03
* :green_heart: [Build #312522](https://buildkite.com/elastic/kibana-pull-request/builds/312522) succeeded eca01bb641386b49e9670aeffc113fa7b98f1eb3
* :broken_heart: [Build #312385](https://buildkite.com/elastic/kibana-pull-request/builds/312385) failed 197f629e3872c479966b184268aae0160a040545
* :broken_heart: [Build #312202](https://buildkite.com/elastic/kibana-pull-request/builds/312202) failed a51b4d6857ffd9f9ec6b59bfb902edfb30c9af4f

cc @maximpn

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/313863","number":313863,"commit":"a86dea3a8b6182f0693015cd201b103af387d2ed"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312777","number":312777,"commit":"341917740e36cea21953f7735a564d11cc7fd134"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312676","number":312676,"commit":"f6cdc52acc443524ac119cb65798d6b344892a03"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312522","number":312522,"commit":"eca01bb641386b49e9670aeffc113fa7b98f1eb3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312385","number":312385,"commit":"197f629e3872c479966b184268aae0160a040545"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312202","number":312202,"commit":"a51b4d6857ffd9f9ec6b59bfb902edfb30c9af4f"}],"number":313863}
buildkite-pr-comment-->

xcrzx commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_management/prebuilt_rules/export_prebuilt_rule.cy.ts:
```diff
@@ -6,214 +6,338 @@
-        index: getIndexPatterns(),
-      });
+    describe('multiple rules', () => {
+      const [PREBUILT_RULE_ID_A, PREBUILT_RULE_ID_B, PREBUILT_RULE_ID_C, PREBUILT_RULE_ID_D] = [
+        'test-prebuilt-rule-a',
+        'test-prebuilt-rule-b',
+        'test-prebuilt-rule-c',
+        'test-prebuilt-rule-d',
+      ];
```
nit: Normal constant declaration is way more readable:

```ts
  const PREBUILT_RULE_ID_A = 'prebuilt-rule-a';
  const PREBUILT_RULE_ID_B = 'prebuilt-rule-b';
  const PREBUILT_RULE_ID_C = 'prebuilt-rule-c';
  const PREBUILT_RULE_ID_D = 'prebuilt-rule-d';
  ```

xcrzx commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/common/import_export/export_prebuilt_rules.ts:
```diff
@@ -0,0 +1,502 @@
+              is_customized: false,
+            },
+          }),
+          expect.objectContaining({
+            rule_id: 'rule-id-1',
+            immutable: false,
+            rule_source: {
+              type: 'internal',
+            },
+          }),
```
Let's use constants instead of hardcoded rule IDs. Something like `rule_id: CUSTOM_RULE_ID` reads much better and makes it effortless to see how the rule was created by searching by references.

xcrzx commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/common/import_export/export_prebuilt_rules.ts:
```diff
@@ -0,0 +1,502 @@
+              is_customized: false,
+            },
+          }),
+          expect.objectContaining({
+            rule_id: PREBUILT_RULE_ID_B,
+            immutable: true,
+            rule_source: {
+              type: 'external',
+              is_customized: false,
+            },
+          }),
```
There's already one non-customized rule in the array, let's add one customized here and remove the `exports customized prebuilt rules via the _export API` scenario altogether.

xcrzx commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/common/import_export/export_prebuilt_rules.ts:
```diff
@@ -0,0 +1,502 @@
+    }),
+    createRuleAssetSavedObject({
+      rule_id: PREBUILT_RULE_ID_B,
+      tags: ['test-tag-2'],
```
Why do we need to define tags here? Are they important for the tests?

xcrzx commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/common/import_export/export_prebuilt_rules.ts:
```diff
@@ -0,0 +1,502 @@
+    beforeEach(async () => {
+      await deleteAllPrebuiltRuleAssets(es, log);
+      await deleteAllRules(supertest, log);
+      await createPrebuiltRuleAssetSavedObjects(es, PREBUILT_RULE_ASSETS);
+      await installPrebuiltRules(es, supertest);
```
Let's move the asset creation and installation that's important for test scenarios into the test body (inside each `it`). Having part of the state initialized in `beforeEach` and the rest in the test body makes it very hard to read and understand; I constantly need to scroll up and down 500 LoC just to figure out the initial state.

It's fine to do cleanup in the `beforeEach` section, but initializing important state there is an anti-pattern in my opinion.

maximpn commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/common/import_export/export_prebuilt_rules.ts:
```diff
@@ -0,0 +1,502 @@
+  const PREBUILT_RULE_ASSETS = [PREBUILT_RULE_A, PREBUILT_RULE_B];
+
+  /**
+   * This test suite is skipped in Serverless MKI environments due to reliance on the
+   * feature flag for prebuilt rule customization.
```
I'd prefer to keep these test skipped in MKI since tests setup requires creating prebuilt rule assets via direct accessing to the system indices. We could overcome it via installing a mock prebuilt rules package but it's less straightforward. This is definitely a separate scope to unskip prebuilt rules related tests in MKI.

maximpn commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_management/prebuilt_rules/export_prebuilt_rule.cy.ts:
```diff
@@ -6,214 +6,338 @@
-        cy.wait('@bulk_action').then(({ response }) => {
-          cy.wrap(response?.body).should('eql', expectedExportedRule(this.customRuleResponse));
-          cy.get(SUCCESS_TOASTER_BODY).should('have.text', 'Successfully exported 1 of 1 rule.');
+      it('exports a mix of non-customized prebuilt, customized prebuilt and custom rules in bulk', () => {
```
Either of these scenarios may go wrong. I'd refrain from having a single test covering all scenarios and looking like a snapshot test.

maximpn commented:
Hey @xcrzx,

Thanks for your review 🙏  I've addressed your comments. I answered in https://github.com/elastic/kibana/pull/224611#discussion_r2179419297 regarding significantly shrinking the number of tests. Could you have a look?

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/314819)
* Commit: 4da2d9ba7889d77348c27d89f2c23449d0b2ca5a

### Failed CI Steps
* [Linting](https://buildkite.com/elastic/kibana-pull-request/builds/314819#0197caa7-62f2-49e8-865a-9d55d1343727)

### History
* :green_heart: [Build #313863](https://buildkite.com/elastic/kibana-pull-request/builds/313863) succeeded a86dea3a8b6182f0693015cd201b103af387d2ed
* :green_heart: [Build #312777](https://buildkite.com/elastic/kibana-pull-request/builds/312777) succeeded 341917740e36cea21953f7735a564d11cc7fd134
* :broken_heart: [Build #312676](https://buildkite.com/elastic/kibana-pull-request/builds/312676) failed f6cdc52acc443524ac119cb65798d6b344892a03
* :green_heart: [Build #312522](https://buildkite.com/elastic/kibana-pull-request/builds/312522) succeeded eca01bb641386b49e9670aeffc113fa7b98f1eb3
* :broken_heart: [Build #312385](https://buildkite.com/elastic/kibana-pull-request/builds/312385) failed 197f629e3872c479966b184268aae0160a040545
* :broken_heart: [Build #312202](https://buildkite.com/elastic/kibana-pull-request/builds/312202) failed a51b4d6857ffd9f9ec6b59bfb902edfb30c9af4f

cc @maximpn

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/314819","number":314819,"commit":"4da2d9ba7889d77348c27d89f2c23449d0b2ca5a"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/313863","number":313863,"commit":"a86dea3a8b6182f0693015cd201b103af387d2ed"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312777","number":312777,"commit":"341917740e36cea21953f7735a564d11cc7fd134"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312676","number":312676,"commit":"f6cdc52acc443524ac119cb65798d6b344892a03"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312522","number":312522,"commit":"eca01bb641386b49e9670aeffc113fa7b98f1eb3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312385","number":312385,"commit":"197f629e3872c479966b184268aae0160a040545"}],"number":314819}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/315042)
* Commit: 3757b7f5551b683a2f36e89a1d0e19da45984384

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #314873](https://buildkite.com/elastic/kibana-pull-request/builds/314873) failed fbcccb2ccc1e2de99324c2e00d1dcb0bc8371c64
* :broken_heart: [Build #314819](https://buildkite.com/elastic/kibana-pull-request/builds/314819) failed 4da2d9ba7889d77348c27d89f2c23449d0b2ca5a
* :green_heart: [Build #313863](https://buildkite.com/elastic/kibana-pull-request/builds/313863) succeeded a86dea3a8b6182f0693015cd201b103af387d2ed
* :green_heart: [Build #312777](https://buildkite.com/elastic/kibana-pull-request/builds/312777) succeeded 341917740e36cea21953f7735a564d11cc7fd134
* :broken_heart: [Build #312676](https://buildkite.com/elastic/kibana-pull-request/builds/312676) failed f6cdc52acc443524ac119cb65798d6b344892a03

cc @maximpn

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/315042","number":315042,"commit":"3757b7f5551b683a2f36e89a1d0e19da45984384"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/314873","number":314873,"commit":"fbcccb2ccc1e2de99324c2e00d1dcb0bc8371c64"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/314819","number":314819,"commit":"4da2d9ba7889d77348c27d89f2c23449d0b2ca5a"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/313863","number":313863,"commit":"a86dea3a8b6182f0693015cd201b103af387d2ed"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312777","number":312777,"commit":"341917740e36cea21953f7735a564d11cc7fd134"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/312676","number":312676,"commit":"f6cdc52acc443524ac119cb65798d6b344892a03"}],"number":315042}
buildkite-pr-comment-->

xcrzx commented on x-pack/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/common/import_export/export_prebuilt_rules.ts:
```diff
@@ -0,0 +1,502 @@
+      );
+    });
+
+    it('exports custom and prebuilt rules via the _export API', async () => {
```
Discussed the changes offline, I'm fine if we keep the tests separated.

xcrzx reviewed (approved):
Thanks for addressing my feedback, Maxim. Went through the changes once again, all seems good 👍

maximpn commented on x-pack/test/security_solution_cypress/cypress/e2e/detection_response/rule_management/prebuilt_rules/export_prebuilt_rule.cy.ts:
```diff
@@ -6,214 +6,338 @@
   },
   () => {
-    describe('Rule export workflow with single rules', () => {
+    before(() => {
+      installMockPrebuiltRulesPackage();
+    });
+
+    beforeEach(() => {
+      deletePrebuiltRulesAssets();
+      deleteAlertsAndRules();
```
Discussed this question offline. Installing a mock package with higher version prevents real package installation. It works much better than `preventPrebuiltRulesPackageInstallation` since the latter simply block `bootstrap` API endpoint requests.

kibanamachine commented:
Starting backport for target branches: 8.18, 8.19, 9.0, 9.1

https://github.com/elastic/kibana/actions/runs/16080334039

kibanamachine commented:
## 💔 Some backports could not be created

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.18|**Backport failed because of merge conflicts**<br><br>You might need to backport the following PRs to 8.18:<br> - [[Security Solution] Implement prebuilt rules import test plan (#223421)](https://github.com/elastic/kibana/pull/223421)|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/226662">](https://github.com/elastic/kibana/pull/226662)|
|❌|9.0|**Backport failed because of merge conflicts**<br><br>You might need to backport the following PRs to 9.0:<br> - [[Security Solution] Implement prebuilt rules import test plan (#223421)](https://github.com/elastic/kibana/pull/223421)|
|✅|9.1|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/226663">](https://github.com/elastic/kibana/pull/226663)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 224611
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->

maximpn commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|9.0|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/226666">](https://github.com/elastic/kibana/pull/226666)|
|✅|8.18|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/226679">](https://github.com/elastic/kibana/pull/226679)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 10.0.0 -->
